qiskit/quantum_info/states/stabilizerstate.py

```diff
@@ -0,0 +1,557 @@
+
+"""
+Stabilizer state class.
+"""
+
+import copy
+import numpy as np
+
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.instruction import Instruction
+from qiskit.exceptions import QiskitError
+from qiskit.quantum_info.states.quantum_state import QuantumState
+from qiskit.quantum_info.operators.symplectic import Clifford, Pauli
+from qiskit.quantum_info.operators.symplectic.clifford_circuits import _append_x
+
+
+class StabilizerState(QuantumState):
+    """StabilizerState class. Based on the internal class:
+    :class:`~qiskit.quantum_info.Clifford`
+
+    References:
+        1. S. Aaronson, D. Gottesman, *Improved Simulation of Stabilizer Circuits*,
+           Phys. Rev. A 70, 052328 (2004).
+           `arXiv:quant-ph/0406196 <https://arxiv.org/abs/quant-ph/0406196>`_
+    """
+
+    def __init__(self, data):
+        """Initialize a StabilizerState object.
+
+        Args:
+            data (StabilizerState or Clifford or Pauli or QuantumCircuit or
+                  qiskit.circuit.Instruction):
+                Data from which the stabilizer state can be constructed.
+
+        Raises:
+            QiskitError: if input data is not a valid stabilizer state (not a valid Clifford).
+        """
+
+        # Initialize from a Clifford
+        if isinstance(data, Clifford):
+            self._data = data
+
+        # Initialize from a Pauli
+        elif isinstance(data, Pauli):
+            self._data = Clifford(data.to_instruction())
+
+        # Initialize from another StabilizerState
+        elif isinstance(data, StabilizerState):
+            self._data = data._data
+
+        # Initialize from a QuantumCircuit or Instruction object
+        elif isinstance(data, (QuantumCircuit, Instruction)):
+            self._data = Clifford.from_circuit(data)
+
+        # Validate that Clifford table is a symplectic matrix
+        if not self.is_valid():
+            raise QiskitError(
+                'Invalid StabilizerState. Input is not a valid Clifford.')
+
+        # Initialize
+        super().__init__()
+
+    def __eq__(self, other):
+        return self._data == other._data
+
+    def __repr__(self):
+        return 'StabilizerState({})'.format(self._data)
+
+    @property
+    def data(self):
+        """Return StabilizerState data"""
+        return self._data
+
+    def is_valid(self, atol=None, rtol=None):
+        """Return True if a valid StabilizerState."""
+        return Clifford.is_unitary(self.data)
+
+    def _add(self, other):
+        raise NotImplementedError(
+            "{} does not support addition".format(type(self)))
+
+    def _multiply(self, other):
+        raise NotImplementedError(
+            "{} does not support scalar multiplication".format(type(self)))
+
+    def trace(self):
+        """Return the trace of the stabilizer state as a density matrix,
+        which equals to 1, since it is always a pure state."""
+        return 1
```

Comment:
Why do we have `if_valid` if we assume we are always valid here?

Reply:
I added `self.is_valid` before returning 1.0.